packages/platform/src/PlatformConfigProvider.ts

```diff
@@ -0,0 +1,113 @@
+import * as Effect from "effect/Effect"
+import * as Either from "effect/Either"
+import * as HashSet from "effect/HashSet"
+import * as Layer from "effect/Layer"
+import * as ReadonlyArray from "effect/ReadonlyArray"
+import { isPlatformError, type PlatformError } from "./Error.js"
+import * as FileSystem from "./FileSystem.js"
+import * as Path from "./Path.js"
+
+/**
+ * @since 1.0.0
+ * @category constructors
+ */
+export const fromFileTree = (options?: {
+  readonly rootDirectory?: string
+}): Effect.Effect<ConfigProvider.ConfigProvider, never, Path.Path | FileSystem.FileSystem> =>
+  Effect.Do.pipe(
+    Effect.bind("path", () => Path.Path),
+    Effect.bind("fs", () => FileSystem.FileSystem),
+    Effect.map(({ fs, path }) => {
+      const rootDirectory = options?.rootDirectory ?? "/"
+
+      const parseConfig = <A>(primitive: Config.Config.Primitive<A>) => (value: string) =>
+        Either.map(primitive.parse(value.trim()), ReadonlyArray.of)
+
+      const readConfig = <A>(filePath: string, primitive: Config.Config.Primitive<A>) =>
+        Effect.flatMap(
+          fs.readFileString(filePath),
+          parseConfig(primitive)
+        )
+
+      const resolveEnumerableDirs = (segments: ReadonlyArray<string>) =>
+        segments.length === 0 ? [] : [path.join(...segments)]
+
+      const resolveFilePath = (pathSegments: ReadonlyArray<string>) => path.join(rootDirectory, ...pathSegments)
+
+      const sourceError = (pathSegments: ReadonlyArray<string>, error: PlatformError) =>
+        ConfigError.SourceUnavailable(
+          [...pathSegments],
+          error.message,
+          Cause.fail(error)
+        )
+      const pathNotFoundError = (pathSegments: ReadonlyArray<string>) =>
+        ConfigError.MissingData(
+          [...pathSegments],
+          `Path ${resolveFilePath(pathSegments)} not found`
+        )
+      const handlePlatformError = (pathSegments: ReadonlyArray<string>) => (error: PlatformError) =>
+        error._tag === "SystemError" && error.reason === "NotFound"
+          ? Effect.fail(pathNotFoundError(pathSegments))
+          : Effect.fail(sourceError(pathSegments, error))
+
+      return ConfigProvider.fromFlat(
+        ConfigProvider.makeFlat({
+          load: (pathSegments, config) =>
+            Effect.catchIf(
+              readConfig(resolveFilePath(pathSegments), config),
+              isPlatformError,
+              handlePlatformError(pathSegments)
+            ),
+          enumerateChildren: (pathSegments) =>
+            Effect.forEach(resolveEnumerableDirs(pathSegments), (dir) => fs.readDirectory(dir)).pipe(
+              Effect.map((files) => HashSet.fromIterable(files.flat())),
+              Effect.catchIf(isPlatformError, handlePlatformError(pathSegments))
+            ),
+          patch: PathPatch.empty
+        })
+      )
+    })
+  )
+
+/**
+ * Add the file tree ConfigProvider to the environment, falling back to the default ConfigProvider.
```

Comment:
Just a nit: I think reading configuration from `process.env` is a more common scenario. If the config is not found there, that's when it should fall back to the file system, not the other way around. Checking if the value exists in `process.env` is also faster than checking the file system for the existence of a file.

Reply:
Yeah I think we can do that.